### PR TITLE
More robust phone number formatting

### PR DESCRIPTION
**Pull request by rkuykendall**
## Phone number formatting

1. Recognize more phone numbers formats
2. Never hide data you don't understand or recognize

image

# Other
- Create mostly internal `canReplaceSymbols`, `replaceSymbolsWithChars` formatting helpers
- Create `hasStringContent` helper that correctly identifies `'  '` as empty
- Create `hasStringOrNumberContent` that correctly identifies `0` as non-empty

### CODE REVIEW

File: src/formatting.tsx [hunk 1]
```diff
@@ -19,21 +21,39 @@ import {
...
+  const charsReverse = chars.reverse();
+
+  return template
+    .split('')
+    .map(char => (char === '#') ? charsReverse.pop() : char)
+    .join('')
+    ;
+}
+
+export function hasStringContent (value: unknown): value is string {
```
[the remark this answers is not in the document]
Comment (rkuykendall) [reply src/formatting.tsx]: Value is unknown, and if this function returns true we KNOW it's a string.

This is great when you have a very lose definition going in like `value?: string | null` but want to do string-specific work. This functions tells typescript that if it returns true, we know it's a string.

File: src/formatting.tsx [hunk 2]
```diff
@@ -46,22 +66,33 @@ export function formatFullName (firstName?: string, lastName?: string) {
...
-  return EMPTY_FIELD;
+
+  return input;
 }
 
 export function formatDate (value?: string | null, dateFormat = DATE_FORMATS.date) {
-  return value
-    ? format(value, dateFormat)
-    : EMPTY_FIELD;
+  if (!hasStringContent(value)) { return EMPTY_FIELD; }
```
[the remark this answers is not in the document]
Comment (rkuykendall) [reply src/formatting.tsx]: Not according to the call string for this function. We really should only be passing ISO strings around as dates so I think that's fine that we don't support... whatever that would be, an epoch?

File: src/formatting.tsx [hunk 3]
```diff
@@ -76,57 +107,62 @@ export function getNameOrDefault (obj?: unknown, { field = 'name', defaultValue
   return defaultValue;
 }
 
-export function getOrDefault (obj?: any) {
-  try {
-    return obj.trim() || EMPTY_FIELD;
+export function getOrDefault (value?: any) {
+  const isUndefined = value === undefined
```
[the remark this answers is not in the document]
Comment (rkuykendall) [reply src/formatting.tsx]: They do, but... IDK why we'd use it here? We would only save 1 char.

File: src/formatting.tsx [hunk 4]
```diff
@@ -19,21 +21,39 @@ import {
 
 import { DATE_FORMATS, EMPTY_FIELD } from './constants';
 
+export function canReplaceSymbols (template: string, chars: string[]): boolean {
+  return (template.split('#').length - 1) === chars.length;
+}
+
+export function replaceSymbolsWithChars (template: string, chars: string[]) {
```
[the remark this answers is not in the document]
Comment (rkuykendall) [reply src/formatting.tsx]: 🧐
```ts
export function replaceSymbolsWithCars (template: string) {
  return replaceSymbolsWithCars(template, new Array(template.length).fill('🚗'));
}
```

File: src/formatting.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (kbouchard) [reply src/formatting.tsx]: I meant the 3 checks `isUndefined` || `isNull` || `isEmptyString` -> `_.lodashDoesItAllUnknown()`.

But no big deal.

**Pull request merged**